File: 4. Kruskal's MST.cpp
#include <bits/stdc++.h>
// Kruskal's Algorithm for Minimum Spanning Tree (MST)
// Greedy strategy: Always pick the smallest weight edge that does not form a cycle (using DSU).
// Task: Given a connected, undirected graph with weighted edges, find an MST using Kruskal's
// algorithm. Return the sum of the weights of the edges in the MST.

#include <algorithm>
#include <iostream>
#include <numeric>
#include <utility>
#include <vector>

using namespace std;

struct Edge
{
    int u;
    int v;
    int w;
};

struct DSU
{
    vector<int> parent;
    vector<int> rankv;

    explicit DSU(int n) : parent(n), rankv(n, 0) { iota(parent.begin(), parent.end(), 0); }

    int find(int x) { return parent[x] == x ? x : parent[x] = find(parent[x]); }

    bool unite(int a, int b)
    {
        a = find(a);
        b = find(b);
        if (a == b)
            return false;
        if (rankv[a] < rankv[b])
            swap(a, b);
        parent[b] = a;
        if (rankv[a] == rankv[b])
            ++rankv[a];
        return true;
    }
};

long long kruskalMST(int n, vector<Edge>& edges)
{
    sort(edges.begin(), edges.end(), [](const Edge& a, const Edge& b) { return a.w < b.w; });
    DSU dsu(n);
    long long total = 0;
    int taken = 0;
    for (const auto& e : edges)
    {
        if (dsu.unite(e.u, e.v))
        {
            total += e.w;
            if (++taken == n - 1)
                break;
        }
    }
    return total;
}

int main()
{
    int n = 4;
    vector<Edge> edges = {{0, 1, 10}, {0, 2, 6}, {0, 3, 5}, {1, 3, 15}, {2, 3, 4}};
    cout << kruskalMST(n, edges) << '\n';
    return 0;
}

/*
Time Complexity: O(E log E + E α(V)) ≈ O(E log E) — sorting edges dominates (α is inverse
Ackermann). Space Complexity: O(V) — DSU parent/rank arrays.
*/

File: 4. Kruskal's MST.py
"""
Kruskal's Algorithm for Minimum Spanning Tree
Task: Given a connected, undirected graph with weighted edges, find an MST using Kruskal's algorithm.
Return the sum of the weights of the edges in the MST.
Greedy strategy: Always pick the smallest weight edge that does not form a cycle (using DSU).
"""
class DSU:
    def __init__(self, n):
        self.p = list(range(n))
        self.r = [0]*n
    def find(self, x):
        if self.p[x] != x:
            self.p[x] = self.find(self.p[x])
        return self.p[x]
    def unite(self, a, b):
        a = self.find(a); b = self.find(b)
        if a == b:
            return False
        if self.r[a] < self.r[b]:
            a, b = b, a
        self.p[b] = a
        if self.r[a] == self.r[b]:
            self.r[a] += 1
        return True

def kruskal_mst(n, edges):
    edges.sort(key=lambda x: x[2])
    dsu = DSU(n)
    total = 0
    taken = 0
    for u, v, w in edges:
        if dsu.unite(u, v):
            total += w
            taken += 1
            if taken == n-1:
                break
    return total

if __name__ == "__main__":
    n = 4
    edges = [(0,1,10),(0,2,6),(0,3,5),(1,3,15),(2,3,4)]
    print(kruskal_mst(n, edges))  # 19
    # Time Complexity: O(E log E + E α(V)) ≈ O(E log E).
    # Space Complexity: O(V) — DSU.
